src/index.ts

```diff
@@ -306,69 +316,84 @@ QuickSQLite.transactionAsync = (
   };
 
   const commit = () => {
+    if (isFinalized) {
+      throw Error(
+        `Quick SQLite Error: Cannot execute commit on finalized transaction: ${dbName}`
+      );
+    }
     const result = QuickSQLite.execute(dbName, 'COMMIT');
     isFinalized = true;
     return result;
   };
 
   const rollback = () => {
+    if (isFinalized) {
+      throw Error(
+        `Quick SQLite Error: Cannot execute rollback on finalized transaction: ${dbName}`
+      );
+    }
     const result = QuickSQLite.execute(dbName, 'ROLLBACK');
     isFinalized = true;
     return result;
   };
 
-  const tx: PendingTransaction = {
-    start: async () => {
-      try {
-        QuickSQLite.execute(dbName, 'BEGIN TRANSACTION');
-        await callback({
-          commit,
-          execute,
-          executeAsync,
-          rollback,
-        });
-
-        if (!isFinalized) {
-          commit();
-        }
-      } catch (e: any) {
-        if (!isFinalized) {
-          rollback();
-        }
+  return await new Promise((resolve, reject) => {
+    const tx: PendingTransaction = {
+      start: async () => {
+        try {
+          QuickSQLite.execute(dbName, 'BEGIN TRANSACTION');
+          await callback({
+            commit,
+            execute,
+            executeAsync,
+            rollback,
+          });
+
+          if (!isFinalized) {
+            commit();
+          }
 
-        // Do not throw an error, because the transaction is executed with a setImmediate call
-        // This errors are uncatchable
-        // https://stackoverflow.com/questions/51081892/nodejs-asynchronous-exceptions-are-uncatchable
+          resolve();
+        } catch (e) {
+          if (!isFinalized) {
+            try {
+              rollback();
+            } catch (rollbackError) {
+              reject(rollbackError);
+            }
+          }
 
-        // throw e;
-      } finally {
-        locks[dbName].inProgress = false;
-        isFinalized = false;
-        startNextTransaction(dbName);
-      }
-    },
-  };
+          reject(e);
+        } finally {
+          locks[dbName].inProgress = false;
+          isFinalized = false;
+          startNextTransaction(dbName);
+        }
+      },
+    };
 
-  locks[dbName].queue.push(tx);
-  startNextTransaction(dbName);
+    locks[dbName].queue.push(tx);
+    startNextTransaction(dbName);
+  });
 };
 
 const startNextTransaction = (dbName: string) => {
+  if (!locks[dbName]) {
+    throw Error(`Lock not found for db: ${dbName}`);
+  }
+
   if (locks[dbName].inProgress) {
     // Transaction is already in process bail out
     return;
   }
 
-  setImmediate(() => {
-    if (!locks[dbName]) {
-      throw Error(`Lock not found for db: ${dbName}`);
-    }
-
-    if (locks[dbName].queue.length) {
-      locks[dbName].inProgress = true;
-      locks[dbName].queue.shift().start();
-    }
-  });
+  if (locks[dbName].queue.length) {
+    locks[dbName].inProgress = true;
```

Comment:
This is the main fix from the first commit - the main issue was that inProgress was set to true inside the setImmediate function which was to late.